missing.py

```diff
@@ -0,0 +1,46 @@
+# L is a list of elements such as serial numbers
+def missing_elements(L, start, end):
+    if end - start <= 1: 
+        if L[end] - L[start] > 1:
+            yield from range(L[start] + 1, L[end])
+        return
+
+    index = start + (end - start) // 2
+    print("index " + str(index))
+    # is the lower half consecutive?
+    consecutive_low =  L[index] == L[start] + (index - start)
+    if not consecutive_low:
+        yield from missing_elements(L, start, index)
+
+    # is the upper part consecutive?
+    consecutive_high =  L[index] == L[end] - (end - index)
+    if not consecutive_high:
+        yield from missing_elements(L, index, end)
+
+def hasduplicate(serialNolist):
+    serialNolist.sort()
+    print(serialNolist)
+    
+    new_list = sorted(set(serialNolist))
+    dup_list =[]
+    
+    for i in range(len(new_list)):
+            if (serialNolist.count(new_list[i]) > 1 ):
+                dup_list.append(new_list[i])
+    print(dup_list)
+
+# num_list is a list of elements such as serial numbers
+def missing_numbers(num_list):
+      original_list = [x for x in range(num_list[0], num_list[-1] + 1)]
+      num_list = set(num_list)
+      return (list(num_list ^ set(original_list)))
+
+# test the methods to check if they are working
+def main():

```

Comment:
Is your plan for the validator library to extract a list of recordIds and serialNumbers then pass them to this method? The messages are a list of json strings and this takes a list of integers.

Reply:
The plan is for the validator to extract the serial numbers (integers) sent them to the library.

Reply:
That's what I was referring to in the meeting. This code relies on sending a list of numbers. So yes, the higher level code, will need to build a list of number and pass it to this method. We need to have that higher level method in the library as well so maintain the same interface.